sdks/java/harness/src/main/java/org/apache/beam/fn/harness/Caches.java

```diff
@@ -0,0 +1,285 @@
+    return new SubCacheable<K, V>() {
+      @Override
+      public V peek(K key) {
+        return null;
+      }
+
+      @Override
+      public V computeIfAbsent(K key, Function<K, V> loadingFunction) {
+        return loadingFunction.apply(key);
+      }
+
+      @Override
+      public void put(K key, V value) {}
+
+      @Override
+      public void clear() {}
+
+      @Override
+      public void remove(K key) {}
+
+      @Override
+      public Set<K> keys() {
+        return Collections.emptySet();
+      }
+    };
+  }
+
+  /**
+   * Uses the specified {@link PipelineOptions} to configure and return a cache instance based upon
+   * parameters within {@link SdkHarnessOptions}.
+   */
+  public static <K, V> Cache<K, V> fromOptions(PipelineOptions options) {
+    // We specifically use cache2k since it allows for recursive computeIfAbsent calls
+    // preventing deadlock from occurring when a loading function mutates the underlying cache
+    org.cache2k.Cache<Object, Object> cache =
+        Cache2kBuilder.forUnknownTypes()
+            .maximumWeight(
+                options.as(SdkHarnessOptions.class).getMaxCacheMemoryUsageMb() * 1024L * 1024L)
+            .weigher(
+                new Weigher<Object, Object>() {
+                  private final MemoryMeter memoryMeter = MemoryMeter.builder().build();
+
+                  @Override
+                  public int weigh(Object key, Object value) {
+                    long size = memoryMeter.measureDeep(key) + memoryMeter.measureDeep(value);
+                    return size > Integer.MAX_VALUE ? Integer.MAX_VALUE : (int) size;
+                  }
+                })
+            .storeByReference(true)
+            .executor(MoreExecutors.directExecutor())
+            .build();
+
+    return (Cache<K, V>) forCache(cache);
+  }
+
+  /**
+   * Returns a view of a cache that operates on keys with a specified key prefix.
+   *
+   * <p>All lookups, insertions, and removals into the parent {@link Cache} will be prefixed by the
+   * specified prefixes.
+   *
+   * <p>Operations which operate over the entire caches contents such as {@link Cache#clear} only
+   * operate over keys with the specified prefixes.
+   */
+  public static <K, V> Cache<K, V> subCache(
+      Cache<?, ?> cache, Object keyPrefix, Object... additionalKeyPrefix) {
+    if (cache instanceof SubCache) {
+      return new SubCache<>(
+          ((SubCache<?, ?>) cache).cache,
+          ((SubCache<?, ?>) cache).keyPrefix.subKey(keyPrefix, additionalKeyPrefix));
+    } else if (cache instanceof SubCacheable) {
+      return new SubCache<>((SubCacheable<?, ?>) cache, CompositeKey.ROOT.subKey(keyPrefix));
+    }
+    throw new UnsupportedOperationException("An unsupported type of cache was passed.");
+  }
+
+  /** A cache that never evicts any values. */
+  public static <K, V> Cache<K, V> eternal() {
+    // We specifically use cache2k since it allows for recursive computeIfAbsent calls
```

Comment:
I assume this issue been known to happen with the Guava cache implementation?

Reply:
This happens with ConcurrentHashMap, Guava, and Caffeine.

Reply:
fwiw, Caffeine handles this through `AsyncCache` where you do not need a separate thread to execute on. See FAQ's [example](https://github.com/ben-manes/caffeine/wiki/Faq#recursive-computations), which is similar to what Apache Solr does in their usage. It is a simple workaround for these rare cases and makes it more explicit about what is happening, as often recursive computes are surprising and can be error prone. I am not sure if one can expect per-key linearizability with recursive writes, so the plus/minus of this behavior depends on the use-case.

One very minor topic to be aware of is that the `Cache#computeIfAbsent` used here does **not** follow the specification for Map's. In addition, until recently the Map's implementation was broken and caused the cache to be in an invalid state. See this [bug report](https://github.com/cache2k/cache2k/issues/174) where I found these issues for the simple single threaded cases. Since these types of mistakes are sadly common and only ironed by usage, you might want to spend a little extra time validating the behavior and avoiding regressions. Caffeine has had its share of bugs and it takes usage to wrangle them out, so early adopters should be a little more proactive to help the library authors catch their oversights.

Reply:
Guava does allow recursive calls, but it is not a support as a feature. It will detect a simple deadlock/livelock when the caller loads for a key it previously started computing (A->B->A). An out-of-order deadlock could occur in any recursive case, though. All three libraries handle this because they compute entries outside of the hash table lock.

Unless this is an advertised feature then this assumption is dangerous. No library has: Guava discourages it, Caffeine warns against it and describes an explicit approach, and cache2k doesn’t seem to be aware of this possibility (not documented or tested for).

The more I think about this, the more I think you need much more rigorous testing and assurance to rely on this working in production facing code. I don’t think this should be left as is.

Reply:
/cc @cpovirk @lowasser maybe you can advise from the Google Java team’s recommendations for this.

Reply:
Yes, `computeIfAbsent` creating a loop is definitely an issue that could happen if this API was misused. I have needed a chain for the use case that I'm building like X -> Y -> Z and is guaranteed to not create a loop otherwise it will cause an infinite loop outside of the cache as well since I'm effectively caching pages and continuation tokens. This is likely to change though as a different implementation won't have the need for the recursive `computeIfAbsent`.

I'll update the docs in the future (and tests) whether this is needed or not depending on the different implementation.

On a different note, I'm trying to ensure that the cache reduces the number of OOMs and maximizes the amount of cache hits by having one cache that is used and a bunch of sub caches that provide views over them for different key spaces. Our previous implementation used distinct cache objects which required one to partition memory and give each cache a limited amount of memory. There didn't seem to be a good way to have the caches work together so that if one was unused the other could have more memory. Note that all of these caches used only weight based eviction.

Reply:
In this case I would recommend Guava `get(key, callable)` because does support X -> Y-> Z, has been used at scale for known robustness, and it's an unofficially supported behavior. Recursive loads was tested for, fixed, discussed, and relied upon internally (e.g. by Gerrit). Otherwise an explicit lock object like the future example is ideal. I don't think you should rely on accidental behavior as your current implementation does. Note that Guava's `get(key, callable)` is very robust for this, but it's `asMap().compute` methods were added semi-recently and likely not.

I'm not familiar enough but would be happy to discuss over email or a video conf if I can be of help. You might be able to use a weak value for your view caches so that they discard when the entry is evicted by the primary cache. Other times I've used index caches that maintain a secondaryKey => primaryKey mapping for a primaryKey => value cache, so it takes two lookups. The callbacks and/or reference caching can help a lot in orchestrating these complex cases.

Reply:
The swap to use guava cache is in https://github.com/apache/beam/pull/16252

Reply:
Thank you @lukecwik! Sorry if I was a bit of a pain about this. 😄

Reply:
(Sorry for not replying. As a general rule, I endorse doing what Ben says; the details of our own caching library (let alone the others out there) are mostly beyond me.)